services/supervisor/tests/supervisor/main.rs

```diff
@@ -813,3 +867,108 @@ fn test_auditor_normal_workflow() {
+    let api = testkit.api(); // update the API
+    assert!(artifact_exists(&api, &artifact.name));
+}
+
+/// This test emulates a deploy confirmation with 12 validators.
+/// Here we send confirmations by the byzantine majority (2/3+1) validators
+/// and expect deploy to start.
+#[test]
+fn test_multiple_validators_deploy_confirm_byzantine_majority() {
+    let validators_count = 12;
+    let byzantine_majority = (validators_count * 2 / 3) + 1;
+    let mut testkit = testkit_with_inc_service_and_n_validators(validators_count);
+    let artifact = artifact_default();
+    let api = testkit.api();
+    assert!(!artifact_exists(&api, &artifact.name));
+
+    let request_deploy = deploy_request(artifact.clone(), testkit.height().next());
+
+    // Send deploy requests by byzantine majority of validators.
+    let deploy_tx_hashes: Vec<exonum_crypto::Hash> = (0..byzantine_majority)
+        .map(|i| deploy_artifact_manually(&mut testkit, &request_deploy, ValidatorId(i)))
+        .collect();
+
+    testkit.create_block();
+    api.exonum_api().assert_txs_success(&deploy_tx_hashes);
+
+    // Send deploy confirmations by every validator.
+    let deploy_confirmations: Vec<Verified<AnyTx>> = (0..validators_count)
+        .map(|i| deploy_confirmation(&testkit, &request_deploy, ValidatorId(i)))
+        .collect();
+
+    testkit.create_block_with_transactions(deploy_confirmations);
+
+    // Check that artifact is deployed now.
+    let api = testkit.api(); // update the API
+    assert!(artifact_exists(&api, &artifact.name));
+}
+
+/// This test emulates a deploy confirmation with 12 validators.
+/// Here we send confirmations by the byzantine minority (2/3) validators
+/// and expect deploy to not start.
+#[test]
+fn test_multiple_validators_deploy_confirm_byzantine_minority() {
+    let validators_count = 12;
+    let byzantine_minority = validators_count * 2 / 3;
+    let mut testkit = testkit_with_inc_service_and_n_validators(validators_count);
+    let artifact = artifact_default();
+    let api = testkit.api();
+    assert!(!artifact_exists(&api, &artifact.name));
+
+    let request_deploy = deploy_request(artifact.clone(), testkit.height().next());
+
+    // Send deploy requests by byzantine majority of validators.
+    let deploy_tx_hashes: Vec<exonum_crypto::Hash> = (0..byzantine_minority)
+        .map(|i| deploy_artifact_manually(&mut testkit, &request_deploy, ValidatorId(i)))
+        .collect();
+
+    testkit.create_block();
+    api.exonum_api().assert_txs_success(&deploy_tx_hashes);
+
+    // Try to send confirmation. It should fail, since deploy was not approved
+    // and thus not registered.
+    let confirmation = deploy_confirmation(&testkit, &request_deploy, ValidatorId(0));
+
+    testkit.add_tx(confirmation.clone());
+    testkit.create_block();
+
+    let expected_status = Err(ExecutionError {
+        kind: ExecutionErrorKind::Service { code: 5 },
+        description: "Deploy request has not been registered.".into(),
```

Comment:
Nit / slightly off-topic: This description looks unclear. In this test, deploy requests are registered (in the sense they are recorded in the blockchain as successful transactions), but their amount is not sufficient to start confirmations.